### ISSUE TITLE
Self-update from Git repository, update remote branches

### ISSUE THREAD

**Issue by uliska** [wbsoft/frescobaldi#216]
This isn't really a feature request but rather an announcement that I'll be investigating this. And I want my thoughts to be documented and open to discussion, therefore this 'Issue'.

As already known I look forward to opening Frescobaldi up to support Git workflows for working with documents/projects. As this may require quite complex considerations it's maybe a good idea to start with something less complex: Frescobaldi itself.

We can update Frescobaldi with a simple `git pull upstream` or switch between different development versions by `git checkout`, and we can even investigate others' versions with new features through `git remote`.
While all this is already available I think it would be great to integrate it into Frescobaldi.

First and most common use is to check for updates and 'install' them automatically.
For this I will start with making Frescobaldi restart with the currently opened document. This is something I find tedious when working on Frescobaldi code: exit Frescobaldi, save the .py file, start Frescobaldi, open a .ly file to test the difference.

The next step will be to make Frescobaldi aware of all remotes that are registered in the Git repository: check for their branches and suggest pulling stuff. Then I will have a list of all branches (->development versions) of all Frescobaldi contributors I decide to add to the remotes list. And I will be able to simply switch to any of these from within Frescobaldi.
This will encourage collaboration I think. If I have written something I can simply ask someone else to give my contribution a try and feedback.

An optional further idea could be (if it's possible to interface that way with Github) to check for pull requests on Github.
I imagine a pop-up on program startup telling me sth. like:
"xxx has opened a pull request on branch new-feature against wbsoft/master. Would you like to check this out? Or view the full pull request on Github?"

**Comment by uliska** [comment]
What has been implemented from this:

* `Git` menu listing all available local branches
* Use that menu to switch to another branch
* Restart action

Technically this builds on a new `vcs` module that wraps around the concept of a (Git) repo. An abstract base class is available, but only Git is actually supported so far.

What has *not* been implemented so far is *anything* that requires interaction with remote repositories. The basic reason for that is my failure to handle conditions that need interaction, for example when the user has configured Git or the repository in a way that requires them to enter a password upon connection. While it is probably possible to interact with a process through the `job.Job` implementation I was strongly advised *not* to try that (https://stackoverflow.com/questions/51599374/pass-root-privilege-to-os-commands-in-python).

There is also much work on an unmerged branch coming from the 2017 GSoC :-(

**Comment by fedelibre** [comment]
"Self-update from Git repository" means that it should pull latest changes of a branch when you switch to that Frescobaldi branch? (I see that now it does only a checkout.)
I guess the difficult part is handling conflicts between local and remote branches, but a checkout shares the same problem, right?

For the current Git menu, only meant to switch Frescobaldi branches, I wonder if current state should be documented.. I see an empty `git.md` file in the usermanual... Let me know if you are interested in a doc patch or you'd rather wait for other features and development before writing docs.

I see in pull #974 that the VCS feature will be documented in [vcs.md](https://github.com/wbsoft/frescobaldi/pull/974/files#diff-ec6d0a433ca90b0082dd06c5cdb38ba0R1).

**Comment by uliska** [comment]
> "Self-update from Git repository" means that it should pull latest changes of a branch when you switch to that Frescobaldi branch? (I see that now it does only a checkout.)

No, but that would be a possibility as well. What I meant is that upon startup Frescobaldi would check if there are updates on the remote and notify the user with the option of pulling and restarting (at least when the current branch has updates).

> I guess the difficult part is handling conflicts between local and remote branches, but a checkout shares the same problem, right?

The *difficult* part is handling the case when the user has to interact to provide credentials, which seems not to be reachable for us. The part about handling conflicts is less "difficult" than "potentially tedious" and requires some extra care.

---

I admit I neglected documenting this because it's (was) a developer-only feature. However, if we want to encourage *users* to also run Frescobalid from Git you are right that it should be documented. Do you have a suggestion how/where this should be placed in the manual, and how the documentation of *this* relates to future documentation of Git features for *document* editing?

**Comment by fedelibre** [comment]
> I admit I neglected documenting this because it's (was) a developer-only feature. However, if we want to encourage _users_ to also run Frescobalid from Git you are right that it should be documented. Do you have a suggestion how/where this should be placed in the manual, and how the documentation of _this_ relates to future documentation of Git features for _document_ editing?

I guess we want to encourage *some* users to test new features before merging them in master (as it happened in the recent past). 
We have a lot of guides in the wiki, which is a good thing (especially considering that the website is currently out-of-date), but I don't think that the user manual should lag behind. User manuals are also translated using proper tools.

We may create a section called *Contributing*, right after Troubleshooting.
Running Frescobaldi from source should be highlighted as a tool for developers as well as users interested in testing features and giving feedback.

About the Git features for document editing, these should be documented in a separate section, perhaps within the *Managing Files* section. Once #974 is merged.

**Comment by fedelibre** [comment]
> What has _not_ been implemented so far is _anything_ that requires interaction with remote repositories. The basic reason for that is my failure to handle conditions that need interaction, for example when the user has configured Git or the repository in a way that requires them to enter a password upon connection. While it is probably possible to interact with a process through the `job.Job` implementation I was strongly advised _not_ to try that (https://stackoverflow.com/questions/51599374/pass-root-privilege-to-os-commands-in-python).

I don't understand why you'd need root privilege to handle git operations. I think above link was related with adding new fonts to lilypond installations in /usr and /usr/local.

**Comment by fedelibre** [comment]
> The _difficult_ part is handling the case when the user has to interact to provide credentials, which seems not to be reachable for us.

Why it's not reachable for us?

Don't know if it can help, but for example [gnome-terminal](https://gitlab.gnome.org/GNOME/gnome-terminal) uses `openssh-askpass` to open a dialog where you can enter remote repository username and password (depending on how `.git/config` is set).

**Comment by uliska** [comment]
> I don't understand why you'd need root privilege to handle git operations. I think above link was related with adding new fonts to lilypond installations in /usr and /usr/local.

You don't need root privileges but you have to supply passwords or other credentials. And I was told to consider this a rabbit hole I wouldn't want to go into. Especially considering the need to support various operating systems. I don't know about that, and I don't want to treat security lightly (by making *me* implement security related functions ...). 

So with some guidance (or even better: contributions) I'd be more than happy to have that added to the feature list. Especially since remote interaction would also be very useful for working with user documents in repositories.

**Comment by fedelibre** [comment]
> We may create a section called _Contributing_, right after Troubleshooting.
> Running Frescobaldi from source should be highlighted as a tool for developers as well as users interested in testing features and giving feedback.

Thinking twice, I wouldn't call it _Contributing_, as this is covered in the README files and shouldn't be doubled in the user manual.
But _Test new features_ may be a valid topic for the user manual. I would link to the wiki pages for explanations on how to run Frescobaldi from source. I don't know how many users run Frescobaldi from source, but I think that the Git menu should be documented and in general encouraging feedback from users is a good thing.

What do you think?
